File: adapter/harness/process_concept.py
"""
Process concept data model and task-graph seeding — P-PC.1, P-PC.2.

A ProcessConcept is a human-authored canonical decomposition for a task class.
It seeds the TaskGraph at harness initialisation and is a strong prior, not
a constraint (INV-PC-01). The model-driven replanning path remains intact.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

# Module-level mtime cache: path (str) → (mtime: float, concept: ProcessConcept)
_FILE_CACHE: dict[str, tuple[float, "ProcessConcept"]] = {}


# ── Errors ────────────────────────────────────────────────────────────────────


class ProcessConceptValidationError(ValueError):
    """Raised when a concept file fails structural validation."""


class ProcessConceptNotFoundError(KeyError):
    """Raised when a requested concept ID is not registered."""


# ── Abstraction level mapping ─────────────────────────────────────────────────

_ABSTRACTION_MAP: dict[str, int] = {
    "module": 0,
    "goal": 0,
    "subgoal": 1,
    "function": 1,
    "leaf": 2,
    "statement": 2,
}

_VALID_RISK_LEVELS = {"LOW", "MEDIUM", "HIGH"}
_VALID_ABSTRACTION_LEVELS = set(_ABSTRACTION_MAP.keys())


# ── Data model ────────────────────────────────────────────────────────────────


@dataclass
class ProcessConceptStep:
    id: str
    description: str
    depends_on: list[str] = field(default_factory=list)
    risk_level: str = "LOW"
    abstraction_level: str = "module"
    expected_tools: list[str] = field(default_factory=list)
    success_criteria: list[str] = field(default_factory=list)
    strategy_hint: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "description": self.description,
            "depends_on": list(self.depends_on),
            "risk_level": self.risk_level,
            "abstraction_level": self.abstraction_level,
            "expected_tools": list(self.expected_tools),
            "success_criteria": list(self.success_criteria),
            "strategy_hint": self.strategy_hint,
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> ProcessConceptStep:
        return cls(
            id=d["id"],
            description=d["description"],
            depends_on=d.get("depends_on", []),
            risk_level=d.get("risk_level", "LOW"),
            abstraction_level=d.get("abstraction_level", "module"),
            expected_tools=d.get("expected_tools", []),
            success_criteria=d.get("success_criteria", []),
            strategy_hint=d.get("strategy_hint"),
        )


@dataclass
class ProcessConcept:
    id: str
    name: str
    description: str
    success_criteria: list[str] = field(default_factory=list)
    schema_version: str = "1"
    steps: list[ProcessConceptStep] = field(default_factory=list)

    # ── Serialisation ──────────────────────────────────────────────────────────

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "success_criteria": list(self.success_criteria),
            "steps": [s.to_dict() for s in self.steps],
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> ProcessConcept:
        return cls(
            id=d["id"],
            name=d.get("name", d["id"]),
            description=d.get("description", ""),
            success_criteria=d.get("success_criteria", []),
            schema_version=d.get("schema_version", "1"),
            steps=[ProcessConceptStep.from_dict(s) for s in d.get("steps", [])],
        )

    @classmethod
    def from_file(cls, path: str | Path) -> ProcessConcept:
        """Load a ProcessConcept from a JSON file, with mtime-keyed caching.

        Re-reads the file only when the mtime has changed since the last load.
        Raises ProcessConceptNotFoundError when the file does not exist.
        Raises ProcessConceptValidationError for invalid content.
        """
        path = Path(path)
        if not path.exists():
            raise ProcessConceptNotFoundError(str(path))

        mtime = path.stat().st_mtime
        cached = _FILE_CACHE.get(str(path))
        if cached is not None and cached[0] == mtime:
            return cached[1]

        try:
            raw = json.loads(path.read_text(encoding="utf-8"))
        except json.JSONDecodeError as exc:
            raise ProcessConceptValidationError(f"Invalid JSON in {path}: {exc}") from exc

        try:
            concept = cls.from_dict(raw)
        except (KeyError, TypeError) as exc:
            raise ProcessConceptValidationError(f"Malformed concept in {path}: {exc}") from exc

        errors = concept.validate()
        if errors:
            raise ProcessConceptValidationError(
                f"Concept {path} failed validation: " + "; ".join(errors)
            )

        _FILE_CACHE[str(path)] = (mtime, concept)
        return concept

    # ── Validation ────────────────────────────────────────────────────────────

    def validate(self) -> list[str]:
        """Return validation errors (empty list = valid).

        Checks: (1) non-empty concept ID, (2) step ID uniqueness,
        (3) depends_on references only known step IDs,
        (4) valid risk_level and abstraction_level values,
        (5) dependency cycle detection.
        """
        errors: list[str] = []

        if not self.id or not self.id.strip():
            errors.append("Concept 'id' must be a non-empty string")

        step_ids: list[str] = []
        seen: set[str] = set()
        for step in self.steps:
            if not step.id or not step.id.strip():
                errors.append("Each step must have a non-empty 'id'")
                continue
            if step.id in seen:
                errors.append(f"Duplicate step id {step.id!r}")
            else:
                seen.add(step.id)
                step_ids.append(step.id)

            if step.risk_level not in _VALID_RISK_LEVELS:
                errors.append(
                    f"Step {step.id!r}: risk_level {step.risk_level!r} must be one of {sorted(_VALID_RISK_LEVELS)}"
                )
            if step.abstraction_level not in _VALID_ABSTRACTION_LEVELS:
                errors.append(
                    f"Step {step.id!r}: abstraction_level {step.abstraction_level!r} "
                    f"must be one of {sorted(_VALID_ABSTRACTION_LEVELS)}"
                )

        id_set = set(step_ids)
        for step in self.steps:
            for dep in step.depends_on:
                if dep not in id_set:
                    errors.append(
                        f"Step {step.id!r} depends_on unknown step {dep!r}"
                    )

        # Cycle detection (iterative DFS)
        adj: dict[str, list[str]] = {s.id: list(s.depends_on) for s in self.steps}
        WHITE, GRAY, BLACK = 0, 1, 2
        colour: dict[str, int] = {sid: WHITE for sid in id_set}
        cycle_reported: set[str] = set()

        def _dfs(start: str) -> bool:
            stack = [(start, iter(adj.get(start, [])))]
            colour[start] = GRAY
            while stack:
                node, children = stack[-1]
                try:
                    child = next(children)
                    if child not in colour:
                        continue
                    if colour[child] == GRAY:
                        return True
                    if colour[child] == WHITE:
                        colour[child] = GRAY
                        stack.append((child, iter(adj.get(child, []))))
                except StopIteration:
                    colour[node] = BLACK
                    stack.pop()
            return False

        for sid in list(id_set):
            if colour[sid] == WHITE:
                if _dfs(sid):
                    if sid not in cycle_reported:
                        errors.append(f"Dependency cycle detected involving step {sid!r}")
                        cycle_reported.add(sid)

        return errors

    # ── TaskGraph seeding ─────────────────────────────────────────────────────

    def seed_task_graph(self, task_graph: Any) -> None:
        """Append concept steps as Task objects to task_graph in-place.

        Task IDs are namespaced as ``{concept_id}:{step_id}`` to avoid
        collisions with model-generated tasks. depends_on references are
        also namespaced. abstraction_level strings are mapped to ints:
        "module"/"goal"→0, "subgoal"/"function"→1, "leaf"/"statement"→2.
        """
        from .task_graph import Task

        for step in self.steps:
            namespaced_id = f"{self.id}:{step.id}"
            namespaced_deps = [f"{self.id}:{dep}" for dep in step.depends_on]
            abs_level = _ABSTRACTION_MAP.get(step.abstraction_level, 0)

            task = Task(
                id=namespaced_id,
                description=step.description,
                status="PENDING",
                depends_on=namespaced_deps,
                risk_level=step.risk_level,  # type: ignore[arg-type]
                assigned_strategy=step.strategy_hint,
                parallel_write_domains=[],
                abstraction_level=abs_level,
                block_reason=None,
                completed_evidence=[],
            )
            task_graph.tasks.append(task)

        task_graph.changed = True

File: packages/harness/src/process-concept.ts
import { type Task, type TaskGraph } from './state/task-graph.js'

export class ProcessConceptValidationError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ProcessConceptValidationError'
  }
}

export class ProcessConceptNotFoundError extends Error {
  constructor(conceptId: string) {
    super(`Process concept "${conceptId}" not found in registry`)
    this.name = 'ProcessConceptNotFoundError'
  }
}

export type RiskLevelStr = 'LOW' | 'MEDIUM' | 'HIGH'
export type AbstractionLevelStr = 'module' | 'goal' | 'subgoal' | 'function' | 'leaf' | 'statement'

export interface ProcessConceptStep {
  id: string
  description: string
  dependsOn: string[]
  riskLevel: RiskLevelStr
  abstractionLevel: AbstractionLevelStr
  expectedTools: string[]
  successCriteria: string[]
  strategyHint: string | null
}

const ABSTRACTION_MAP: Record<AbstractionLevelStr, number> = {
  module: 0,
  goal: 0,
  subgoal: 1,
  function: 1,
  leaf: 2,
  statement: 2,
}

const VALID_RISK_LEVELS = new Set<RiskLevelStr>(['LOW', 'MEDIUM', 'HIGH'])
const VALID_ABSTRACTION_LEVELS = new Set<AbstractionLevelStr>([
  'module', 'goal', 'subgoal', 'function', 'leaf', 'statement',
])

export class ProcessConcept {
  id: string
  name: string
  description: string
  successCriteria: string[]
  schemaVersion: string
  steps: ProcessConceptStep[]

  constructor(data: {
    id: string
    name: string
    description: string
    successCriteria: string[]
    schemaVersion: string
    steps: ProcessConceptStep[]
  }) {
    this.id = data.id
    this.name = data.name
    this.description = data.description
    this.successCriteria = data.successCriteria
    this.schemaVersion = data.schemaVersion
    this.steps = data.steps
  }

  validate(): string[] {
    const errors: string[] = []

    if (!this.id || !this.id.trim()) {
      errors.push("Concept 'id' must be a non-empty string")
    }

    const stepIds: string[] = []
    const seen = new Set<string>()

    for (const step of this.steps) {
      if (!step.id || !step.id.trim()) {
        errors.push('Each step must have a non-empty id')
        continue
      }
      if (seen.has(step.id)) {
        errors.push(`Duplicate step id "${step.id}"`)
      } else {
        seen.add(step.id)
        stepIds.push(step.id)
      }

      if (!VALID_RISK_LEVELS.has(step.riskLevel)) {
        errors.push(`Step "${step.id}": riskLevel "${step.riskLevel}" must be one of HIGH, LOW, MEDIUM`)
      }
      if (!VALID_ABSTRACTION_LEVELS.has(step.abstractionLevel)) {
        errors.push(
          `Step "${step.id}": abstractionLevel "${step.abstractionLevel}" must be one of function, goal, leaf, module, statement, subgoal`,
        )
      }
    }

    const idSet = new Set(stepIds)
    for (const step of this.steps) {
      for (const dep of step.dependsOn) {
        if (!idSet.has(dep)) {
          errors.push(`Step "${step.id}" depends_on unknown step "${dep}"`)
        }
      }
    }

    // Cycle detection via iterative DFS
    const adj: Record<string, string[]> = {}
    for (const step of this.steps) {
      if (idSet.has(step.id)) adj[step.id] = step.dependsOn.filter(d => idSet.has(d))
    }
    const WHITE = 0, GRAY = 1, BLACK = 2
    const colour: Record<string, number> = {}
    for (const id of stepIds) colour[id] = WHITE

    const hasCycle = (start: string): boolean => {
      const stack: [string, number, string[]][] = [[start, 0, adj[start] ?? []]]
      colour[start] = GRAY
      while (stack.length > 0) {
        const frame = stack[stack.length - 1]
        const [node, idx, children] = frame
        if (idx >= children.length) {
          colour[node] = BLACK
          stack.pop()
          continue
        }
        frame[1]++
        const child = children[idx]
        if (!(child in colour)) continue
        if (colour[child] === GRAY) return true
        if (colour[child] === WHITE) {
          colour[child] = GRAY
          stack.push([child, 0, adj[child] ?? []])
        }
      }
      return false
    }

    for (const id of stepIds) {
      if (colour[id] === WHITE) {
        if (hasCycle(id)) {
          errors.push(`Dependency cycle detected involving step "${id}"`)
        }
      }
    }

    return errors
  }

  static fromJson(json: Record<string, unknown>): ProcessConcept {
    if (!json.id) {
      throw new ProcessConceptValidationError("Concept JSON missing required field 'id'")
    }

    const rawSteps = (json.steps as Record<string, unknown>[] | undefined) ?? []
    const steps: ProcessConceptStep[] = rawSteps.map((s, i) => {
      const step = s as Record<string, unknown>
      if (!step.id) {
        throw new ProcessConceptValidationError(`Step at index ${i} missing required field 'id'`)
      }
      const riskLevel = (step.risk_level ?? step.riskLevel ?? 'LOW') as string
      const abstractionLevel = (step.abstraction_level ?? step.abstractionLevel ?? 'module') as string
      if (!VALID_RISK_LEVELS.has(riskLevel as RiskLevelStr)) {
        throw new ProcessConceptValidationError(
          `Step "${step.id}": invalid riskLevel "${riskLevel}"`,
        )
      }
      return {
        id: String(step.id),
        description: String(step.description ?? ''),
        dependsOn: ((step.depends_on ?? step.dependsOn ?? []) as string[]).map(String),
        riskLevel: riskLevel as RiskLevelStr,
        abstractionLevel: abstractionLevel as AbstractionLevelStr,
        expectedTools: ((step.expected_tools ?? step.expectedTools ?? []) as string[]).map(String),
        successCriteria: ((step.success_criteria ?? step.successCriteria ?? []) as string[]).map(String),
        strategyHint: (step.strategy_hint ?? step.strategyHint ?? null) as string | null,
      }
    })

    const concept = new ProcessConcept({
      id: String(json.id),
      name: String(json.name ?? json.id),
      description: String(json.description ?? ''),
      successCriteria: ((json.success_criteria ?? json.successCriteria ?? []) as string[]).map(String),
      schemaVersion: String(json.schema_version ?? json.schemaVersion ?? '1'),
      steps,
    })

    const errors = concept.validate()
    if (errors.length > 0) {
      throw new ProcessConceptValidationError(`Concept validation failed: ${errors.join('; ')}`)
    }

    return concept
  }

  seedTaskGraph(taskGraph: TaskGraph): void {
    for (const step of this.steps) {
      const task: Task = {
        id: `${this.id}:${step.id}`,
        description: step.description,
        status: 'PENDING',
        risk_level: step.riskLevel,
        depends_on: step.dependsOn.map(dep => `${this.id}:${dep}`),
        parallel_write_domains: [],
        abstraction_level: ABSTRACTION_MAP[step.abstractionLevel] ?? 0,
        assigned_strategy: step.strategyHint,
      }
      taskGraph.tasks.push(task)
    }
    taskGraph.changed = true
  }
}
